src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets

```diff
@@ -220,24 +220,31 @@ Copyright (c) .NET Foundation. All rights reserved.
     </RebuildDependsOn>
   </PropertyGroup>
 
-  <Target Name="_AddMicrosoftNetCompilerToolsetFrameworkPackage" Condition="'$(MSBuildRuntimeType)' == 'Full'" BeforeTargets="CollectPackageReferences">
+  <PropertyGroup Condition="'$(MSBuildRuntimeType)' == 'Full'">
+    <!-- Automatically opt users into using the toolset package if they are running an MSBuild other than what this SDK was built against.
+          This is to reduce 'tearing'/dependency mismatch, but as always users can override this behavior by disabling the hosted compiler flag. -->
+    <BuildWithNetFrameworkHostedCompiler Condition="'$(BuildWithNetFrameworkHostedCompiler)' == '' and '$(_IsDisjointMSBuildVersion)' == 'true'">true</BuildWithNetFrameworkHostedCompiler>
+  </PropertyGroup>
 
-    <!-- Users should not be setting Microsoft.Net.Compilers.Toolset.Framework directly.
-         If they do, and they also set BuildWithNetFrameworkHostedCompiler, we will have
-         a duplicate PackageReference. This makes it more explicit that that is not supported. -->
-    <NETSdkWarning ResourceName="CannotDirectlyReferenceMicrosoftNetCompilersToolsetFramework"
-        Condition="'@(PackageReference->AnyHaveMetadataValue('Identity', 'Microsoft.Net.Compilers.Toolset.Framework'))' == 'true'" />
+  <PropertyGroup Condition="'$(BuildWithNetFrameworkHostedCompiler)' == 'true' and '$(OS)' == 'Windows_NT'">
+    <RoslynTargetsPath>$(NuGetPackageRoot)\microsoft.net.sdk.compilers.toolset\$(NETCoreSdkVersion)</RoslynTargetsPath>
```

Comment:
I think for `PackageDownload`, `GeneratePathProperty` is automatically set to true.  Any reason not to use the property NuGet generates for the package?

Also, what happens during the restore phase when the package hasn't been downloaded yet.  Is it OK for restore if this points to a non-existent path?

Reply:
> I think for `PackageDownload`, `GeneratePathProperty` is automatically set to true. Any reason not to use the property NuGet generates for the package?

I think the property is not available for PackageDownload: https://github.com/NuGet/Home/issues/8476

Reply:
> Also, what happens during the restore phase when the package hasn't been downloaded yet. Is it OK for restore if this points to a non-existent path?

I think yes, the compiler shouldn't be used during or before restore, right?
Also setting the path after restoring doesn't seem to have any effect, I guess the UsingTask is considered sooner and cannot be overwritten?

Reply:
> Also, what happens during the restore phase when the package hasn't been downloaded yet. Is it OK for restore if this points to a non-existent path?

Looks like you were right, that is a problem in Visual Studio: https://github.com/dotnet/sdk/pull/42238